File: algorithms/_test/paradigmas-bit-manipulation.py
# Bit Manipulation

# # Encontrar elemento único (XOR)
def single_number(nums: list[int]) -> int:
    """
    Encontra o elemento que aparece apenas uma vez em uma lista onde todos os outros
    aparecem exatamente duas vezes, usando XOR.
    
    Exemplo:
        >>> single_number([2, 1, 4, 5, 2, 4, 1])
        5
    """
    result = 0
    for x in nums:
        result ^= x
    return result

if __name__ == "__main__":
    print(single_number([2, 1, 4, 5, 2, 4, 1]))  # Saída: 5


# # Contagem de bits
def count_bits(n: int) -> int:
    """
    Conta quantos bits '1' existem na representação binária de n,
    usando o algoritmo de Brian Kernighan (remove o último bit 1 a cada iteração).
    
    Exemplo:
        >>> count_bits(13)  # 13 = 0b1101
        3
    """
    count = 0
    while n:
        n &= n - 1
        count += 1
    return count

if __name__ == "__main__":
    print(count_bits(13))  # Saída: 3
    print(count_bits(0b101010))  # Saída: 3


# # Subsets via máscara de bits
def subsets(nums: list[int]) -> list[list[int]]:
    """
    Gera todos os subconjuntos de nums usando máscara de bits.
    
    Exemplo:
        >>> subsets([1, 2, 3])
        [[], [1], [2], [1,2], [3], [1,3], [2,3], [1,2,3]]
    """
    n = len(nums)
    res = []
    for mask in range(1 << n):
        subset = []
        for i in range(n):
            if mask & (1 << i):
                subset.append(nums[i])
        res.append(subset)
    return res

if __name__ == "__main__":
    print(subsets([1, 2, 3]))
    # Saída:
    # [[], [1], [2], [1,2], [3], [1,3], [2,3], [1,2,3]]

File: algorithms/_test/paradigmas-bit-manipulation.rb
# Bit Manipulation em Ruby

# Encontra o elemento que aparece uma vez (xor)
def single_number(nums)
  result = 0
  nums.each { |x| result ^= x }
  result
end

# Conta bits '1' usando Brian Kernighan
def count_bits(n)
  count = 0
  while n != 0
    n &= (n - 1)
    count += 1
  end
  count
end

# Gera todos os subconjuntos via máscara de bits
def subsets(nums)
  n = nums.size
  res = []
  (0...1<<n).each do |mask|
    subset = []
    nums.each_with_index do |num, i|
      subset << num if (mask & (1 << i)) != 0
    end
    res << subset
  end
  res
end

# Exemplos de uso
if __FILE__ == $PROGRAM_NAME
  p single_number([2, 1, 4, 5, 2, 4, 1])   # => 5
  p count_bits(13)                         # => 3
  p count_bits(0b101010)                   # => 3
  p subsets([1, 2, 3])                     # => [[], [1], [2], [1,2], [3], [1,3], [2,3], [1,2,3]]
end
